source/module_basis/module_ao/parallel_2d.cpp

```diff
@@ -1,209 +1,127 @@
-    ncol = numroc_(&N_A, &nb, &coord[1], &zero, &dim1);
-    nloc = nrow * ncol;
+    nrow = numroc_(&mg, &nb, &coord[0], &zero, &dim0);
+    ncol = numroc_(&ng, &nb, &coord[1], &zero, &dim1);
+    nloc = static_cast<int64_t>(nrow) * ncol;
 
-    // mohan add 2010-09-12
-    if (nrow == 0 || ncol == 0)
-    {
-        ofs_warning << " cpu 2D distribution : " << dim[0] << "*" << dim[1] << std::endl;
-        ofs_warning << " but, the number of row and column blocks are "
-                    << M_A / nb + (M_A % nb != 0) << "*" << N_A / nb + (N_A % nb != 0)
-                    << std::endl;
-        if (nb > 1)
-        {
-            return 1;
-        }
-        else
-        {
-            ModuleBase::WARNING_QUIT("Parallel_2D::set_local2global", "some processor has no blocks, try a smaller 'nb2d' parameter or reduce the number of mpi processes.");
-        }
-    }
+    // initialize the ScaLAPACK descriptor
+    int info = 0, lld = std::max(nrow, 1);
+    descinit_(desc, &mg, &ng, &nb, &nb, &zero, &zero, &blacs_ctxt, &lld, &info);
 
+    // generate the global-to-local and local-to-global index maps
     local2global_row_.resize(nrow);
+    global2local_row_ = std::vector<int>(mg, -1);
     for (int i = 0; i < nrow; ++i)
     {
-        local2global_row_[i] = (i / nb) * dim0 * nb + coord[0] * nb + i % nb;
+        local2global_row_[i] = (i / nb * dim0 + coord[0]) * nb + i % nb;
+        global2local_row_[local2global_row_[i]] = i;
     }
 
     local2global_col_.resize(ncol);
+    global2local_col_ = std::vector<int>(ng, -1);
     for (int j = 0; j < ncol; ++j)
     {
-        local2global_col_[j] = (j / nb) * dim1 * nb + coord[1] * nb + j % nb;
+        local2global_col_[j] = (j / nb * dim1 + coord[1]) * nb + j % nb;
+        global2local_col_[local2global_col_[j]] = j;
     }
+}
 
-    return 0;
+
+int Parallel_2D::init(
+    const int mg,
+    const int ng,
+    const int nb,
+    const MPI_Comm comm,
+    const bool mode
+)
+{
+    _init_proc_grid(comm, mode);
+    _set_dist_info(mg, ng, nb);
+    return nrow == 0 || ncol == 0;
 }
-#else
-void Parallel_2D::set_serial(const int& M_A, const int& N_A)
+
+
+int Parallel_2D::set(
+    const int mg,
+    const int ng,
+    const int nb,
+    const MPI_Comm comm_2D,
+    const int blacs_ctxt
+)
 {
-    ModuleBase::TITLE("Parallel_2D", "set_serial");
-    this->nrow = M_A;
-    this->ncol = N_A;
-    this->nloc = this->nrow * this->ncol;
-    this->local2global_row_.resize(this->nrow);
-    this->local2global_col_.resize(this->ncol);
-    for (int i = 0; i < this->nrow; i++) this->local2global_row_[i] = i;
-    for (int i = 0; i < this->ncol; i++) this->local2global_col_[i] = i;
+    this->comm_2D = comm_2D;
+    this->blacs_ctxt = blacs_ctxt;
+    Cblacs_gridinfo(blacs_ctxt, &dim0, &dim1, &coord[0], &coord[1]);
+    _set_dist_info(mg, ng, nb);
+    return nrow == 0 || ncol == 0;
 }
 #endif
+
+void Parallel_2D::set_serial(const int mg, const int ng)
+{
+    nb = 1;
```

Comment:
Would it be better to assert(mg>0) and assert(ng>0)?

Reply:
An assertion has been added in the latest commit.